File: codes/python/wk03_sho.py
"""
Week 3 - Quantum Harmonic Oscillator
Physical Chemistry 2, SKKU (Prof. S. Joon Kwon)

Solves the SHO via Hermite polynomials:
  psi_n(y) = (1/pi)^(1/4) / sqrt(2^n n!) * H_n(y) * exp(-y^2/2),
  y = sqrt(m*omega/hbar) x,  E_n = (n + 1/2) hbar*omega
Verifies: Hermite orthogonality, normalization, virial theorem,
and the classically forbidden ("tunneling") probability.
"""
import numpy as np
from math import factorial, pi, sqrt

trapz = getattr(np, "trapezoid", getattr(np, "trapz", None))  # numpy 1.x/2.x compat

def hermite(n, y):
    """Physicists' Hermite polynomial H_n(y) via H_{n+1} = 2y H_n - 2n H_{n-1}."""
    if n == 0: return np.ones_like(y)
    if n == 1: return 2.0 * y
    Hm, Hc = np.ones_like(y), 2.0 * y
    for k in range(1, n):
        Hm, Hc = Hc, 2.0 * y * Hc - 2.0 * k * Hm
    return Hc

def psi(n, y):
    """Normalized SHO eigenfunction in dimensionless y (hbar = m = omega = 1)."""
    norm = 1.0 / sqrt(2.0**n * factorial(n)) * (1.0 / pi)**0.25
    return norm * hermite(n, y) * np.exp(-y**2 / 2.0)

if __name__ == "__main__":
    y = np.linspace(-12, 12, 240001)

    # 1) Hermite orthogonality: int e^{-y^2} H_m H_n dy = 2^n n! sqrt(pi) delta_mn
    print("Hermite orthogonality (integral / 2^n n! sqrt(pi)):")
    for m in range(4):
        row = [trapz(np.exp(-y**2) * hermite(m, y) * hermite(n, y), y)
               / (2.0**n * factorial(n) * sqrt(pi)) for n in range(4)]
        print("  ", ["%+.6f" % v for v in row])

    # 2) Normalization of psi_n
    for n in [0, 1, 5, 10]:
        print(f"<psi_{n}|psi_{n}> = {trapz(psi(n, y)**2, y):.6f}")

    # 3) Virial theorem: <V> = E_n/2 (units of hbar*omega)
    for n in [0, 3]:
        V = trapz(0.5 * y**2 * psi(n, y)**2, y)
        print(f"n={n}: <V> = {V:.5f}  (E_n/2 = {(n + 0.5) / 2:.5f})")

    # 4) Classical turning point y_tp = sqrt(2n+1) & forbidden-region probability
    for n in [0, 1, 5]:
        ytp = sqrt(2 * n + 1.0)
        Pout = trapz(np.where(np.abs(y) > ytp, psi(n, y)**2, 0.0), y)
        print(f"n={n}: y_tp = {ytp:.4f}, P(classically forbidden) = {Pout:.4f}")

File: codes/cpp/wk03_sho.cpp
// Week 3 - Quantum Harmonic Oscillator (Hermite polynomials)
// Physical Chemistry 2, SKKU (Prof. S. Joon Kwon)
// psi_n(y) = (1/pi)^(1/4)/sqrt(2^n n!) H_n(y) exp(-y^2/2), E_n = (n+1/2) hbar*w
// Build: g++ -O2 -std=c++17 wk03_sho.cpp -o wk03_sho
#include <cstdio>
#include <cmath>
#include <vector>

double hermiteH(int n, double y) {           // H_{n+1} = 2y H_n - 2n H_{n-1}
    if (n == 0) return 1.0;
    if (n == 1) return 2.0 * y;
    double Hm = 1.0, Hc = 2.0 * y;
    for (int k = 1; k < n; ++k) {
        double Hn = 2.0 * y * Hc - 2.0 * k * Hm;
        Hm = Hc; Hc = Hn;
    }
    return Hc;
}

double factorial(int n) { double f = 1; for (int i = 2; i <= n; ++i) f *= i; return f; }

double psiSHO(int n, double y) {             // hbar = m = omega = 1
    double norm = 1.0 / std::sqrt(std::pow(2.0, n) * factorial(n)) * std::pow(M_PI, -0.25);
    return norm * hermiteH(n, y) * std::exp(-y * y / 2.0);
}

// trapezoid rule on a uniform grid
template <typename F>
double integrate(F f, double a, double b, int N) {
    double h = (b - a) / N, s = 0.5 * (f(a) + f(b));
    for (int i = 1; i < N; ++i) s += f(a + i * h);
    return s * h;
}

int main() {
    const double A = 12.0; const int NG = 240000;

    // 1) Hermite orthogonality: int e^{-y^2} H_m H_n dy = 2^n n! sqrt(pi) delta_mn
    std::printf("Hermite orthogonality (normalized):\n");
    for (int m = 0; m < 4; ++m) {
        for (int n = 0; n < 4; ++n) {
            double I = integrate([&](double y) {
                return std::exp(-y * y) * hermiteH(m, y) * hermiteH(n, y); }, -A, A, NG);
            std::printf(" %+.6f", I / (std::pow(2.0, n) * factorial(n) * std::sqrt(M_PI)));
        }
        std::printf("\n");
    }

    // 2) Normalization & virial theorem
    for (int n : {0, 1, 5, 10}) {
        double I = integrate([&](double y) { double p = psiSHO(n, y); return p * p; }, -A, A, NG);
        std::printf("<psi_%d|psi_%d> = %.6f\n", n, n, I);
    }
    for (int n : {0, 3}) {
        double V = integrate([&](double y) { double p = psiSHO(n, y); return 0.5 * y * y * p * p; }, -A, A, NG);
        std::printf("n=%d: <V> = %.5f (E_n/2 = %.5f)\n", n, V, (n + 0.5) / 2.0);
    }

    // 3) Classically forbidden probability beyond y_tp = sqrt(2n+1)
    for (int n : {0, 1, 5}) {
        double ytp = std::sqrt(2.0 * n + 1.0);
        double P = 2.0 * integrate([&](double y) { double p = psiSHO(n, y); return p * p; }, ytp, A, NG);
        std::printf("n=%d: y_tp = %.4f, P(forbidden) = %.4f\n", n, ytp, P);
    }
    return 0;
}
